src/main/groovy/org/ajoberstar/grgit/auth/JschAgentProxySessionFactory.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2012-2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ajoberstar.grgit.auth;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import org.eclipse.jgit.transport.JschConfigSessionFactory;
+import org.eclipse.jgit.transport.OpenSshConfig.Host;
+import org.eclipse.jgit.util.FS;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.jcraft.jsch.IdentityRepository;
+import com.jcraft.jsch.JSch;
+import com.jcraft.jsch.JSchException;
+import com.jcraft.jsch.Session;
+import com.jcraft.jsch.agentproxy.Connector;
+import com.jcraft.jsch.agentproxy.RemoteIdentityRepository;
+import com.jcraft.jsch.agentproxy.USocketFactory;
+import com.jcraft.jsch.agentproxy.connector.PageantConnector;
+import com.jcraft.jsch.agentproxy.connector.SSHAgentConnector;
+import com.jcraft.jsch.agentproxy.usocket.JNAUSocketFactory;
+import com.jcraft.jsch.agentproxy.usocket.NCUSocketFactory;
+
+/**
+ * A session factory that supports use of ssh-agent and Pageant SSH authentication.
+ *
+ * @since 0.1.0
+ */
+public class JschAgentProxySessionFactory extends JschConfigSessionFactory {
+  private static final Logger logger = LoggerFactory.getLogger(JschAgentProxySessionFactory.class);
+  private final AuthConfig config;
+
+  public JschAgentProxySessionFactory(AuthConfig config) {
+    this.config = config;
+  }
+
+  /** Customize session */
+  @Override
+  protected void configure(Host hc, Session session) {
+    config.getSessionConfig().forEach(session::setConfig);
+  }
+
+  /**
+   * Obtains a JSch used for creating sessions, with the addition of ssh-agent and Pageant agents,
+   * if available.
+   *
+   * @return the JSch instance
+   */
+  @Override
+  protected JSch getJSch(Host hc, FS fs) throws JSchException {
+    JSch jsch;
+    try {
+      jsch = super.getJSch(hc, fs);
+    } catch (JSchException e) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1166)